### ISSUE TITLE
Multiplayer rejoin issue

### ISSUE THREAD

**Issue by DThomas44** [leehuk/factorio-tnfp#5]
Players can initially join server without issue.

If a player disconnects, occasionally gets error message of:
> Cannot join. The following mod script files are not identical between you and the server:
> mod-TrainNetworkForPlayers

No other information provided to the player, no information in the server's log file.
Can see player download map in log file, but player is disconnected from server.

> 39430.968 ConnectionRequestReplyConfirm (type(ConnectionRequestReplyConfirm) ) from(IP ADDR:({192.168.2.203:55012}))
> 39430.968 Refusing connection for address (IP ADDR:({192.168.2.203:55012})), username (Dinenon). UserVerificationMissing
> 39431.104 ConnectionRequestReplyConfirm (type(ConnectionRequestReplyConfirm) ) from(IP ADDR:({192.168.2.203:55013}))
> 39431.104 Info ServerRouter.cpp:423: Replying to connectionRequest for address(IP ADDR:({192.168.2.203:55013})).
> 39431.104 Info ServerSynchronizer.cpp:600: nextHeartbeatSequenceNumber(1198265) adding peer(6)
> 39431.128 Info ServerMultiplayerManager.cpp:776: updateTick(35684906) changing state from(InGame) to(InGameSavingMap)
> 39431.129 Info ServerMultiplayerManager.cpp:925: updateTick(35684906) received stateChanged peerID(6) oldState(Ready) newState(ConnectedWaitingForMap)
> 39433.876 Info ServerMultiplayerManager.cpp:993: UpdateTick(35684906) Serving map(/opt/factorio/temp/mp-save-5.zip) for peer(6) size(31835480) crc(2149907575)
> 39433.876 Info ServerMultiplayerManager.cpp:776: updateTick(35684906) changing state from(InGameSavingMap) to(InGame)
> 39433.944 Info ServerMultiplayerManager.cpp:925: updateTick(35684906) received stateChanged peerID(6) oldState(ConnectedWaitingForMap) newState(ConnectedDownloadingMap)
> 39439.025 Info ServerMultiplayerManager.cpp:925: updateTick(35684906) received stateChanged peerID(6) oldState(ConnectedDownloadingMap) newState(ConnectedLoadingMap)
> 39465.204 Info ServerMultiplayerManager.cpp:1039: Disconnect notification for peer (6)
> 39465.204 Info ServerSynchronizer.cpp:619: nextHeartbeatSequenceNumber(1199420) removing peer(6).
> 39466.200 Info ServerMultiplayerManager.cpp:1189: Auto saving map as /opt/factorio/saves/mmi.zip
> 39466.201 Info AppManager.cpp:394: Saving game as /opt/factorio/saves/mmi.zip
> 39469.013 Info AppManagerStates.cpp:1754: Saving finished

Any player that receives this error cannot rejoin until the server is restarted.

Unsure if user closing client between connection attempts affects, testing inconclusive.
Unsure if time between disconnecting and reconnecting affects issue, testing inconclusive.

**Comment by leehuk** [comment]
I've been taking a look at this today but at this stage I'm not quite sure entirely whats going on other than theres some form of desync issue.  This is (unluckily for you!) quite possibly the first real test of it running on a multiplayer instance and it clearly hasn't worked how I expected it to.

I've managed to get a headless instance running now though which means I can actually test it to some degree and through that managed to reproduce the same desync issue, but I think its likely to need at least a couple of weeks to dig into whats going on before I can see about whats involved with making multiplayer actually work (and actually testing it).

Sadly that means you're going to be best off not using the TNfP mod in multiplayer for now, but I'll keep this issue updated as I work through it.

**Comment by DThomas44** [comment]
No worries, only a small group (2-4) players on my server, and they are all admins, so any of us can restart as needed. Other than that the mod works great so far, will continue posting if we find any further issues.

**Comment by leehuk** [comment]
Always nice to have good feedback! :)

So the good news is I've managed to work out whats going on and its not as bad as I thought to fix thankfully.

Basically we use the stdlib mod for a couple of its math functions, but it turns out that the position module we use for calculating an area to find nearby train stops, actually dynamically includes a separate area module -- this triggers a desync in multiplayer because the area module ends up loaded on the server, but when a client connects it isn't loaded as part of the save.  Restarting the server unloads the area module, allowing clients to connect in again.

I've got a pending fix which basically swaps the stdlib position module out for some simple inbuilt math functions and everything seems to be a lot more stable with that.  There's a couple of other things I've found to take a look at, but I'll look to sort out a v0.4.2 release for you within the next couple of days or so.
